File: src/com/intellij/idea/plugin/hybris/type/system/meta/impl/TSMetaItemImpl.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.type.system.meta.impl;

import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaCustomProperty;
import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaDeployment;
import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaItem;
import com.intellij.idea.plugin.hybris.type.system.meta.impl.CaseInsensitive.NoCaseMultiMap;
import com.intellij.idea.plugin.hybris.type.system.model.Attribute;
import com.intellij.idea.plugin.hybris.type.system.model.CreationMode;
import com.intellij.idea.plugin.hybris.type.system.model.Index;
import com.intellij.idea.plugin.hybris.type.system.model.ItemType;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.util.xml.DomAnchor;
import com.intellij.util.xml.DomService;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by Martin Zdarsky-Jones (martin.zdarsky@hybris.com) on 15/06/2016.
 */
public class TSMetaItemImpl extends TSMetaEntityImpl<ItemType> implements TSMetaItem {

    private final NoCaseMultiMap<TSMetaItemAttribute> myAttributes = new NoCaseMultiMap<>();
    private final NoCaseMultiMap<TSMetaCustomProperty> myCustomProperties = new NoCaseMultiMap<>();
    private final NoCaseMultiMap<TSMetaItemIndex> myIndexes = new NoCaseMultiMap<>();
    private final Set<DomAnchor<ItemType>> myAllDoms = new LinkedHashSet<>();

    private final TSMetaDeployment<TSMetaItem> myDeployment;
    private String myExtendedMetaItemName;
    private final boolean myAbstract;
    private final boolean myAutoCreate;
    private final boolean myGenerate;
    private final boolean mySingleton;
    private final boolean myJaloOnly;
    private final String myJaloClass;
    private final String myDescription;

    @SuppressWarnings("ThisEscapedInObjectConstruction")
    public TSMetaItemImpl(final Module module, final Project project, final String name, final @NotNull ItemType dom, final boolean custom) {
        super(module, project, name, dom, custom);
        myAllDoms.add(DomService.getInstance().createAnchor(dom));
        registerExtends(dom);
        myAbstract = Boolean.TRUE.equals(dom.getAbstract().getValue());
        myAutoCreate = Boolean.TRUE.equals(dom.getAutoCreate().getValue());
        myGenerate = Boolean.TRUE.equals(dom.getGenerate().getValue());
        mySingleton = Boolean.TRUE.equals(dom.getSingleton().getValue());
        myJaloOnly = Boolean.TRUE.equals(dom.getJaloOnly().getValue());
        myJaloClass = dom.getJaloClass().getStringValue();
        myDescription = Optional.ofNullable(dom.getDescription().getXmlTag())
                                .map(description -> description.getValue().getText())
                                .orElse(null);
        myDeployment = new TSMetaDeploymentImpl<>(module, project, this, dom.getDeployment(), custom);
    }

    @NotNull
    @Override
    public Stream<? extends ItemType> retrieveAllDomsStream() {
        return myAllDoms.stream()
                        .map(DomAnchor::retrieveDomElement)
                        .filter(Objects::nonNull);
    }

    @Override
    public void addAttribute(final String key, final TSMetaItemAttribute attribute) {
        myAttributes.putValue(key, attribute);
    }

    @Override
    public void addCustomProperty(final String key, final TSMetaCustomProperty customProperty) {
        myCustomProperties.putValue(key, customProperty);
    }

    @Override
    public void addIndex(final String key, final TSMetaItemIndex index) {
        myIndexes.putValue(key, index);
    }

    @Override
    public NoCaseMultiMap<TSMetaItemAttribute> getAttributes() {
        return myAttributes;
    }

    @Override
    public NoCaseMultiMap<TSMetaCustomProperty> getCustomAttributes() {
        return myCustomProperties;
    }

    @Override
    public NoCaseMultiMap<TSMetaItemIndex> getIndexes() {
        return myIndexes;
    }

    @Nullable
    @Override
    public String getExtendedMetaItemName() {
        return myExtendedMetaItemName;
    }

    @Override
    public void merge(final TSMetaItem another) {
        addDomRepresentation(another.retrieveDom());
    }

    @Override
    public TSMetaDeployment<TSMetaItem> getDeployment() {
        return myDeployment;
    }

    protected void addDomRepresentation(final @NotNull ItemType anotherDom) {
        myAllDoms.add(DomService.getInstance().createAnchor(anotherDom));
        registerExtends(anotherDom);
    }

    private void registerExtends(final @NotNull ItemType dom) {
        //only one extends is allowed
        if (myExtendedMetaItemName == null) {
            myExtendedMetaItemName = dom.getExtends().getRawText();
        }
    }

    @Override
    public boolean isAbstract() {
        return myAbstract;
    }

    @Override
    public boolean isAutoCreate() {
        return myAutoCreate;
    }

    @Override
    public boolean isGenerate() {
        return myGenerate;
    }

    @Override
    public boolean isSingleton() {
        return mySingleton;
    }

    @Override
    public boolean isJaloOnly() {
        return myJaloOnly;
    }

    @Override
    public String getJaloClass() {
        return myJaloClass;
    }

    @Override
    public String getDescription() {
        return myDescription;
    }

    public static class TSMetaItemIndexImpl extends TSMetaEntityImpl<Index> implements TSMetaItemIndex {

        private final TSMetaItem myOwner;
        private final boolean myRemove;
        private final boolean myReplace;
        private final boolean myUnique;
        private final Set<String> myKeys;
        private final CreationMode myCreationMode;

        public TSMetaItemIndexImpl(final Module module, final Project project, final @NotNull TSMetaItem owner, final @NotNull Index dom, final boolean custom) {
            super(module, project, extractName(dom), dom, custom);
            myOwner = owner;
            myRemove = Boolean.TRUE.equals(dom.getRemove().getValue());
            myReplace = Boolean.TRUE.equals(dom.getReplace().getValue());
            myUnique = Boolean.TRUE.equals(dom.getUnique().getValue());
            myCreationMode = Optional.ofNullable(dom.getCreationMode().getValue()).orElse(CreationMode.ALL);
            myKeys = dom.getKeys().stream()
                        .map(indexKey -> indexKey.getAttribute().getStringValue())
                        .collect(Collectors.toSet());
        }

        @Nullable
        @Override
        public String getName() {
            return super.getName();
        }

        @Override
        public Set<String> getKeys() {
            return Collections.unmodifiableSet(myKeys);
        }

        @Nullable
        private static String extractName(final Index dom) {
            return dom.getName().getStringValue();
        }

        @Override
        public boolean isRemove() {
            return myRemove;
        }

        @Override
        public boolean isReplace() {
            return myReplace;
        }

        @Override
        public boolean isUnique() {
            return myUnique;
        }

        @Override
        public CreationMode getCreationMode() {
            return myCreationMode;
        }

        @Override
        public TSMetaItem getOwner() {
            return myOwner;
        }
    }

    public static class TSMetaItemAttributeImpl extends TSMetaEntityImpl<Attribute> implements TSMetaItemAttribute {

        private final TSMetaItem myOwner;
        private final NoCaseMultiMap<TSMetaCustomProperty> myCustomProperties = new NoCaseMultiMap<>();
        private final boolean myDeprecated;
        private final boolean myAutoCreate;
        private final boolean myGenerate;
        private final boolean myRedeclare;
        private final String myDescription;
        private final String myDefaultValue;
        @Nullable private final String myType;

        public TSMetaItemAttributeImpl(final Module module, final Project project, final @NotNull TSMetaItem owner, final @NotNull Attribute dom, final boolean custom) {
            super(module, project, extractName(dom), dom, custom);
            myOwner = owner;
            myDeprecated = extractDeprecated(dom);
            myRedeclare = Boolean.TRUE.equals(dom.getRedeclare().getValue());
            myAutoCreate = Boolean.TRUE.equals(dom.getAutoCreate().getValue());
            myGenerate = Boolean.TRUE.equals(dom.getGenerate().getValue());
            myType = dom.getType().getStringValue();
            myDescription = Optional.ofNullable(dom.getDescription().getXmlTag())
                .map(xmlTag -> xmlTag.getValue().getText())
                .orElse(null);
            myDefaultValue = dom.getDefaultValue().getStringValue();
            dom.getCustomProperties().getProperties().stream()
                    .map(domAttribute -> new TSMetaCustomPropertyImpl(module, project, domAttribute, custom))
                    .filter(attribute -> StringUtils.isNotBlank(attribute.getName()))
                    .forEach(attribute -> addCustomProperty(attribute.getName().trim(), attribute));
        }

        @Override
        @Nullable
        public String getType() {
            return myType;
        }

        @Override
        public boolean isDeprecated() {
            return myDeprecated;
        }

        @Override
        public boolean isAutoCreate() {
            return myAutoCreate;
        }

        @Override
        public boolean isRedeclare() {
            return myRedeclare;
        }

        @Override
        public boolean isGenerate() {
            return myGenerate;
        }

        @Override
        public void addCustomProperty(final String key, final TSMetaCustomProperty customProperty) {
            myCustomProperties.putValue(key, customProperty);
        }

        @Override
        public @NotNull List<? extends TSMetaCustomProperty> getCustomProperties(final boolean includeInherited) {
            return new LinkedList<>(myCustomProperties.values());
        }

        @Nullable
        @Override
        public String getName() {
            return super.getName();
        }

        @Nullable
        @Override
        public String getDescription() {
            return myDescription;
        }

        @Nullable
        @Override
        public String getDefaultValue() {
            return myDefaultValue;
        }

        @NotNull
        @Override
        public TSMetaItem getOwner() {
            return myOwner;
        }

        @Nullable
        private static String extractName(final Attribute dom) {
            return dom.getQualifier().getStringValue();
        }

        private boolean extractDeprecated(@NotNull final Attribute dom) {
            final String name = getName();
            return name != null && dom.getModel().getSetters().stream().anyMatch(
                setter -> name.equals(setter.getName().getStringValue()) &&
                          Boolean.TRUE.equals(setter.getDeprecated().getValue()));
        }

    }
}

File: src/com/intellij/idea/plugin/hybris/system/type/meta/model/impl/TSMetaItemImpl.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for IntelliJ IDEA.
 * Copyright (C) 2019-2025 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.system.type.meta.model.impl

import com.intellij.idea.plugin.hybris.system.type.meta.TSGlobalMetaModel
import com.intellij.idea.plugin.hybris.system.type.meta.TSMetaHelper
import com.intellij.idea.plugin.hybris.system.type.meta.TSMetaModelException
import com.intellij.idea.plugin.hybris.system.type.meta.impl.CaseInsensitive.CaseInsensitiveConcurrentHashMap
import com.intellij.idea.plugin.hybris.system.type.meta.model.*
import com.intellij.idea.plugin.hybris.system.type.meta.model.TSMetaItem.TSMetaItemAttribute
import com.intellij.idea.plugin.hybris.system.type.meta.model.TSMetaItem.TSMetaItemIndex
import com.intellij.idea.plugin.hybris.system.type.model.Attribute
import com.intellij.idea.plugin.hybris.system.type.model.CreationMode
import com.intellij.idea.plugin.hybris.system.type.model.Index
import com.intellij.idea.plugin.hybris.system.type.model.ItemType
import com.intellij.idea.plugin.hybris.util.xml.toBoolean
import com.intellij.util.xml.DomAnchor
import com.intellij.util.xml.DomService
import java.util.*

internal class TSMetaItemImpl(
    dom: ItemType,
    override val moduleName: String,
    override val extensionName: String,
    override val name: String?,
    override var isCustom: Boolean,
    override val attributes: Map<String, TSMetaItemAttribute>,
    override val indexes: Map<String, TSMetaItemIndex>,
    override val customProperties: Map<String, TSMetaCustomProperty>,
    override val deployment: TSMetaDeployment?
) : TSMetaItem {

    override val domAnchor: DomAnchor<ItemType> = DomService.getInstance().createAnchor(dom)
    override val isAbstract = dom.abstract.toBoolean()
    override val isAutoCreate = dom.autoCreate.toBoolean()
    override val isGenerate = dom.generate.toBoolean()
    override val isSingleton = dom.singleton.toBoolean()
    override val isJaloOnly = dom.jaloOnly.toBoolean()
    override val isCatalogAware = TSMetaHelper.isCatalogAware(dom.customProperties)
    override val jaloClass = dom.jaloClass.stringValue
    override val description = dom.description.xmlTag?.value?.text
    override var extendedMetaItemName = dom.extends.stringValue

    override fun toString() = "Item(module=$extensionName, name=$name, isCustom=$isCustom)"

    internal class TSMetaItemIndexImpl(
        dom: Index,
        override val moduleName: String,
        override val extensionName: String,
        override val name: String,
        override var isCustom: Boolean
    ) : TSMetaItemIndex {

        override val domAnchor: DomAnchor<Index> = DomService.getInstance().createAnchor(dom)
        override val isRemove = dom.remove.toBoolean()
        override val isReplace = dom.replace.toBoolean()
        override val isUnique = dom.unique.toBoolean()
        override val creationMode = dom.creationMode.value ?: CreationMode.ALL
        override val keys = dom.keys
            .mapNotNull { it.attribute.stringValue }
            .toSet()
        override val includes = dom.includes
            .mapNotNull { it.attribute.stringValue }
            .toSet()

        override fun toString() = "Index(module=$extensionName, name=$name, isCustom=$isCustom)"
    }

    internal class TSMetaItemAttributeImpl(
        dom: Attribute,
        override val moduleName: String,
        override val extensionName: String,
        override val name: String,
        override var isCustom: Boolean,
        override val persistence: TSMetaPersistence,
        override val modifiers: TSMetaModifiers,
        override val customProperties: Map<String, TSMetaCustomProperty>
    ) : TSMetaItemAttribute {

        override val domAnchor: DomAnchor<Attribute> = DomService.getInstance().createAnchor(dom)

        override val description = dom.description.xmlTag?.value?.text
        override val defaultValue = dom.defaultValue.stringValue
        override val type = dom.type.stringValue
        override val isDeprecated = TSMetaHelper.isDeprecated(dom.model, name)
        override val isAutoCreate = dom.autoCreate.toBoolean()
        override val isGenerate = dom.generate.toBoolean()
        override val isRedeclare = dom.redeclare.toBoolean()
        override val isSelectionOf = dom.isSelectionOf.stringValue
        override val isLocalized = TSMetaHelper.isLocalized(type)
        override val isDynamic = TSMetaHelper.isDynamic(persistence)

        override fun toString() = "Attribute(module=$extensionName, name=$name, isCustom=$isCustom)"
    }
}

internal class TSGlobalMetaItemImpl(localMeta: TSMetaItem) : TSGlobalMetaItemSelfMerge<ItemType, TSMetaItem>(localMeta), TSGlobalMetaItem {

    override val attributes = CaseInsensitiveConcurrentHashMap<String, TSGlobalMetaItem.TSGlobalMetaItemAttribute>()
    override val customProperties = CaseInsensitiveConcurrentHashMap<String, TSMetaCustomProperty>()
    override val indexes = CaseInsensitiveConcurrentHashMap<String, TSGlobalMetaItem.TSGlobalMetaItemIndex>()
    override val relationEnds = LinkedList<TSMetaRelation.TSMetaRelationElement>()

    override val allAttributes = CaseInsensitiveConcurrentHashMap<String, TSGlobalMetaItem.TSGlobalMetaItemAttribute>()
    override val allOrderingAttributes = CaseInsensitiveConcurrentHashMap<String, TSMetaRelation.TSMetaOrderingAttribute>()
    override val allIndexes = LinkedList<TSGlobalMetaItem.TSGlobalMetaItemIndex>()
    override val allCustomProperties = LinkedList<TSMetaCustomProperty>()
    override val allRelationEnds = LinkedList<TSMetaRelation.TSMetaRelationElement>()
    override val allExtends = LinkedHashSet<TSGlobalMetaItem>()

    override var domAnchor = localMeta.domAnchor
    override var moduleName = localMeta.moduleName
    override var extensionName = localMeta.extensionName
    override var extendedMetaItemName = localMeta.extendedMetaItemName
    override var isAbstract = localMeta.isAbstract
    override var isAutoCreate = localMeta.isAutoCreate
    override var isGenerate = localMeta.isGenerate
    override var isSingleton = localMeta.isSingleton
    override var isJaloOnly = localMeta.isJaloOnly
    override var jaloClass = localMeta.jaloClass
    override var description = localMeta.description
    override var deployment = localMeta.deployment
    override var isCatalogAware = localMeta.isCatalogAware
    override var flattenType: String? = TSMetaHelper.flattenType(this)

    init {
        mergeAttributes(localMeta, this)
        mergeIndexes(localMeta)
        mergeCustomProperties(localMeta)
    }

    override fun toString() = "Item(module=$extensionName, name=$name, isCustom=$isCustom)"

    @Suppress("UNCHECKED_CAST")
    private fun mergeAttributes(localMeta: TSMetaItem, globalMeta: TSGlobalMetaItemImpl) = localMeta.attributes.values.forEach {
        val globalAttribute = this.attributes.computeIfAbsent(it.name) { _ -> TSGlobalMetaItemAttributeImpl(globalMeta, it) }
        (globalAttribute as? TSMetaSelfMerge<Attribute, TSMetaItemAttribute>)
            ?.merge(it)
    }

    @Suppress("UNCHECKED_CAST")
    private fun mergeIndexes(localMeta: TSMetaItem) = localMeta.indexes.values.forEach {
        val globalIndex = this.indexes.computeIfAbsent(it.name) { _ -> TSGlobalMetaItemIndexImpl(it) }
        (globalIndex as? TSMetaSelfMerge<Index, TSMetaItemIndex>)?.merge(it)
    }

    private fun mergeCustomProperties(localMeta: TSMetaItem) = customProperties.putAll(localMeta.customProperties)

    override fun postMerge(globalMetaModel: TSGlobalMetaModel) {
        val extends = declarations
            .map { it.extendedMetaItemName }
            .flatMap {
                try {
                    return@flatMap TSMetaHelper.getAllExtends(globalMetaModel, name, it)
                } catch (e: TSMetaModelException) {
                    e.message?.let { message -> this.mergeConflicts.add(message) }

                    return@flatMap emptyList()
                }
            }
            .toSet()
        val currentRelationEnds = TSMetaHelper.getAllRelationEnds(globalMetaModel, this, emptySet())
        val combinedRelationEnds = TSMetaHelper.getAllRelationEnds(globalMetaModel, this, extends)

        allExtends.addAll(extends)
        allAttributes.putAll(attributes)
        extends.forEach { allAttributes.putAll(it.attributes) }
        allCustomProperties.addAll(customProperties.values + extends.flatMap { it.customProperties.values })
        allIndexes.addAll(indexes.values + extends.flatMap { it.indexes.values })
        allRelationEnds.addAll(combinedRelationEnds)
        relationEnds.addAll(currentRelationEnds)

        if (!isCatalogAware) {
            isCatalogAware = extends.any { it.isCatalogAware }
        }
    }

    override fun mergeInternally(localMeta: TSMetaItem) {
        if (localMeta.isAbstract) isAbstract = localMeta.isAbstract
        if (localMeta.isAutoCreate) isAutoCreate = localMeta.isAutoCreate
        if (localMeta.isGenerate) isGenerate = localMeta.isGenerate
        if (localMeta.isJaloOnly) isJaloOnly = localMeta.isJaloOnly
        if (localMeta.isSingleton) isJaloOnly = localMeta.isSingleton
        if (localMeta.isCatalogAware) isCatalogAware = localMeta.isCatalogAware

        if (localMeta.deployment?.retrieveDom()?.exists() == true) deployment = localMeta.deployment

        localMeta.extendedMetaItemName?.let {
            if (extendedMetaItemName != null) mergeConflicts.add("Extends should be defined only once.")
            extendedMetaItemName = it
        }

        mergeAttributes(localMeta, this)
        mergeIndexes(localMeta)
        mergeCustomProperties(localMeta)
    }

    internal class TSGlobalMetaItemIndexImpl(localMeta: TSMetaItemIndex) : TSMetaSelfMerge<Index, TSMetaItemIndex>(localMeta), TSGlobalMetaItem.TSGlobalMetaItemIndex {

        override val name: String = localMeta.name
        override var domAnchor = localMeta.domAnchor
        override var moduleName = localMeta.moduleName
        override var extensionName = localMeta.extensionName
        override var isRemove = localMeta.isRemove
        override var isReplace = localMeta.isReplace
        override var isUnique = localMeta.isUnique
        override var creationMode = localMeta.creationMode
        override var keys = localMeta.keys
        override var includes = localMeta.includes

        override fun mergeInternally(localMeta: TSMetaItemIndex) {
            if (localMeta.isReplace) {
                isRemove = localMeta.isRemove
                isReplace = localMeta.isReplace
                isUnique = localMeta.isUnique
                creationMode = localMeta.creationMode
                keys = localMeta.keys
                includes = localMeta.includes
            }
        }

        override fun toString() = "Index(module=$extensionName, name=$name, isCustom=$isCustom)"
    }

    internal class TSGlobalMetaItemAttributeImpl(
        override val owner: TSGlobalMetaItem,
        localMeta: TSMetaItemAttribute
    ) : TSMetaSelfMerge<Attribute, TSMetaItemAttribute>(localMeta),
        TSGlobalMetaItem.TSGlobalMetaItemAttribute {

        override val customProperties = CaseInsensitiveConcurrentHashMap<String, TSMetaCustomProperty>()
        override val name: String = localMeta.name
        override var moduleName = localMeta.moduleName
        override var extensionName = localMeta.extensionName
        override var modifiers = localMeta.modifiers
        override var persistence = localMeta.persistence
        override var domAnchor = localMeta.domAnchor
        override var description = localMeta.description
        override var defaultValue = localMeta.defaultValue
        override var type = localMeta.type
        override var isDeprecated = localMeta.isDeprecated
        override var isAutoCreate = localMeta.isAutoCreate
        override var isGenerate = localMeta.isGenerate
        override var isRedeclare = localMeta.isRedeclare
        override var isSelectionOf = localMeta.isSelectionOf
        override var isLocalized = localMeta.isLocalized
        override var isDynamic = localMeta.isDynamic

        // type will be flattened after merge, we need to know exact type to expand it
        override var flattenType: String? = null

        init {
            mergeCustomProperties(localMeta)
        }

        private fun mergeCustomProperties(localMeta: TSMetaItemAttribute) = customProperties.putAll(localMeta.customProperties)

        override fun mergeInternally(localMeta: TSMetaItemAttribute) {
            if (localMeta.isRedeclare) {
                domAnchor = localMeta.domAnchor
                type = localMeta.type
                modifiers = localMeta.modifiers
                isLocalized = localMeta.isLocalized
                isDynamic = localMeta.isDynamic

                mergeCustomProperties(localMeta)
            }
        }

        override fun toString() = "Attribute(module=$extensionName, name=$name, isCustom=$isCustom)"

    }
}